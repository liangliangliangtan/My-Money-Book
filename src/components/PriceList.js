/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import Ionicon from 'react-ionicons';
import {addPlusOrminusSign} from '../utilities/PriceUtility';
import PropTypes from 'prop-types'

/**
 * Stateless Component for displaying the items in the List 
 * @param {*} 
 */
const PriceList = ( ({items, onEditItem, onDeleteItem}) => {

    return (
        <ul className= "list-group list-group-flush">
            {
                items.map( (item) => (
                        <li className="list-group-item d-flex 
                                       justify-content-between align-items-center" 
                            key={item.id} >

                            <span className="col-1">
                                <Ionicon    icon={item.category.iconName} 
                                            fontSize="30px" 
                                            color="#fff"
                                            style={{ backgroundColor: '#347eff', padding: '5px'}}>
                                </Ionicon>
                            </span>

                            <span className='col-5'>
                                {item.title}
                            </span>

                            <span className= "col-2 font-weight-bold">
                                {addPlusOrminusSign(item.category.type)} {item.price}
                            </span>

                            <span className="col-2">
                                {item.date}
                            </span>

                            <a  className="col-1" role="button"
                                    style={{cursor: 'pointer'}}
                                    onClick ={ (event) =>{ event.preventDefault(); onEditItem(item)}}>
                                    <Ionicon icon= "md-create"
                                            fontSize="20px" 
                                            color="#7B68EE">
                                    </Ionicon>
                            </a>

                            <a  className="col-1" role="button" 
                                style={{cursor: 'pointer'}}
                                onClick ={ (event) =>{ event.preventDefault(); onDeleteItem(item)}}>
                                    <Ionicon icon= "md-trash"
                                            fontSize="20px" 
                                            color="#A9A9A9">
                                    </Ionicon>
                            </a>
                            
                        </li> 
                    
                ))
            }    
        </ul>
    )
})

PriceList.propTypes ={
    items : PropTypes.array.isRequired,
    onEditItem: PropTypes.func.isRequired,
    onDeleteItem: PropTypes.func.isRequired
}

export default PriceList;