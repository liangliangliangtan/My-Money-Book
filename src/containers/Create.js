
import React from 'react'
//import {Carousels, Carousel} from '../components/Carousels'
//import {images} from '../utilities/imageData'
import {Tabs,Tab} from '../components/Tabs';
import CateogrySelect from '../components/CategorySelect';
import { TYPE_INCOME, TYPE_OUTCOME} from '../utilities/PriceUtility';
import PriceForm from '../components/PriceForm';
import withContext from '../WithContext'
import { withRouter } from 'react-router-dom'

export const tabsArr = [TYPE_OUTCOME, TYPE_INCOME]
class Create extends React.Component{

    constructor(props){
        super(props)
        const { items, categories } = this.props.data
        const { id } = this.props.match.params
        this.state = {
            selectedTab: id && items[id] ? categories[items[id].cid].type: tabsArr[0],
            selectedCategory: id && items[id] ? categories[items[id].cid]: null,
            isValidForm: true 
        }
    }

    /**
     * For refresh or directly visit the create page
     */
    componentDidMount() {
        const { id } = this.props.match.params
        this.props.actions.getEditDataById(id).then((data)=>{
            const {categories, editItem} = data
            if(editItem){
                this.setState({
                    selectedTab: categories[editItem.cid].type,
                    selectedCategory: categories[editItem.cid] 
                })
            }
        })
    }
    
    setSelectedTab = (index)=>{
        this.setState({
            selectedTab:tabsArr[index]
        })
    }

    setSelectedCategory = (category)=>{
        this.setState({
            selectedCategory: category
        })
    }

    /**
     * Cancel submit means that nothing should be done,
     * return back to the home page 
     * 
     */
    cancelsubmit = ()=> {
        this.props.history.push('/')
    }

    /**
     * Submit the form,call the actions defined in the context provider 
     * actions.createItem: if creating an item 
     * actions.editItem: if editing an item
     */
    submitForm = (submitItem, isEdit) => {

        if (!this.state.selectedCategory){
            this.setState({
                isValidForm: false
            })
            return 
        }
        this.setState({
            isValidForm: true
        })

        if(!isEdit){
            this.props.actions.createItem(submitItem, this.state.selectedCategory).then(()=>{
                this.props.history.push('/')
            })
        }else{
            this.props.actions.updateItem(submitItem, this.state.selectedCategory).then(()=>{
                this.props.history.push('/')
            })
        }
        
    }
    

    render(){

        const {items, categories} = this.props.data
        const { selectedCategory, selectedTab} = this.state
        const {id} = this.props.match.params
        const editItem = (id && items[id]) ? items[id]: null
        
        /**
         * Get all the cateogries which is corresponding to the selected Tab
         */
        const categoriesWithType = Object.keys(categories).map((id) => {
                return categories[id]
            }).filter((category) => {
                return category.type === this.state.selectedTab
            })

        const tabIndex = tabsArr.findIndex((text) => (text === selectedTab))

        return(
            
            <div className='create-page py-3 px-3 rounded mt-3' style={{ background: '#fff' }}>
                <Tabs activeIndex={tabIndex} onTabChange={this.setSelectedTab}>
                    <Tab>OutCome</Tab>
                    <Tab>InCome</Tab>
                </Tabs>
                <CateogrySelect categories={categoriesWithType} onSelectCategory={this.setSelectedCategory}
                    selectedCategory={selectedCategory}>
                </CateogrySelect>
                <PriceForm onFormSubmit={this.submitForm} onCancelSubmit={this.cancelsubmit} editItem={editItem}/>
                { !this.state.isValidForm &&ã€€
                    <div className="alert alert-danger col-sm-4 pr-3" role="alert">
                        Category must be selected
                    </div>
                } 
            </div>
        )
    }
}


export default withRouter(withContext(Create))


