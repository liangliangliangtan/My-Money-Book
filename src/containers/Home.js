import React from 'react';
import logo from '../logo.svg';
import PriceList from '../components/PriceList';
import TotalPrice from '../components/TotalPrice';
import CreateButton from '../components/CreateButton';
import { TYPE_INCOME, TYPE_OUTCOME, padZeroToLeft} from '../utilities/PriceUtility';
import { LIST_MODE, CHART_MODE, TREND_MODE} from '../utilities/PriceUtility';
import MonthPicker from '../components/MonthPicker';
import {Tabs,Tab} from'../components/Tabs';
import Ionicon from 'react-ionicons';
import LineFigure from '../components/LineFigure'
import withContext from '../WithContext'
import {withRouter} from 'react-router-dom'
import Loader from '../components/Loader'
import PieFigure from'../components/PieFigure'
import Footer from '../components/Footer'
import {Carousels} from '../components/Carousels'
import {images} from '../utilities/imageData'
export const categories = {
    "1": {
            "id": "1",
            "iconName": "md-restaurant",
            "type": TYPE_OUTCOME
         },

    "2": {
            "id": "2",
            "iconName": "md-train",
            "type": TYPE_INCOME
         }
}

export const items = [
    {
        "id": 1,
        "title": "testItem",
        "price": 200,
        "date": "2018-09-10",
        "categoryId": 1
    },

    {
        "id": 2,
        "title": "testItem2",
        "price": 300,
        "date": "2018-09-10",
        "categoryId" : 2
    }
]

export const tabArray = [LIST_MODE, CHART_MODE, TREND_MODE]

/**
 * Home component is a container contains following display component: 
 * PriceList,Tabs, TotalPrice ,MonthPicker, 
 * 
 * Design State:
 *    1. items List (with categoryId as foreign key constrains )
 *    2. View Mode(List Or Chart )
 */
class Home extends React.Component {
    
    constructor(props){
        super(props)
        this.state = {
            items: this.props.data.items,
            tabMode: tabArray[0]  //LIST_MODE
        }
    }

    componentDidMount() {
        this.props.actions.getInitialData()
    }

    /**
     * Change the view of the tab component according to the currentView passing by the tab component
     */
    changeView = (index) => {
        this.setState({
            tabMode: tabArray[index]
        })
    }

    /**
     * Change the year and month selected by the MonthPicker Component 
     * Notice: 
     */
    changeDate = (selectedYear, selectedMonth) => {
        console.log(selectedYear, selectedMonth)
        this.props.actions.getDataByMonthPicker(selectedYear, selectedMonth)
    }

    /**TODO: Redirect to the Edited page */
    modeifyItem = (modifyItem) => {
        this.props.history.push(`/edit/${modifyItem.id}`)
    }

    /**
     * Jump to the create  page 
     * (Usage: push(path, [state]) - (function) Pushes a new entry onto the history stack)
     * 
     */
    createItem = () => {
        this.props.history.push('/create')
    }

    deleteItem = (itemTobeDeleted) => {
        this.props.actions.deleteItem(itemTobeDeleted)
    }

    render(){

        const { items, categories, currentDate, isLoading, headLines} = this.props.data       
        const {tabMode} = this.state
        console.log(headLines)
        /**
         * Mapping items with their belonging category by categoryId
         * and filter those items within the selected year and month
         */
        const itemswithCategory = Object.keys(items).map((id) =>{
            let curItem = items[id]
            curItem.category = categories[curItem.cid];
            return curItem
        }).filter( (curItem) => {
            return curItem.date.includes(`${currentDate.year}-${padZeroToLeft(currentDate.month)}`)
        })

        let totalIncome =0, totalOutcome = 0;
        // compute the total prices
        itemswithCategory.forEach((item) => {
            if (item.category.type === TYPE_OUTCOME){
                totalOutcome += item.price 
            } else if (item.category.type ===  TYPE_INCOME){
                totalIncome += item.price
            }
        })

        // filter items by their types 
        const filterItemsByType = (items, type) =>{
            return items.filter((item) => {
                return item.category.type === type
            })
        }

        // convert from tab mode text to tab index 
        const tabIndex = tabArray.findIndex(text => text === tabMode);
            
            return (
                <React.Fragment>
                        <header className="App-Header">
                        {   headLines && headLines.length !== 0 &&
                            <Carousels height={360} width={'75%'} totalItems={headLines.length} items={headLines}>
                            </Carousels>
                        }
                            <div className="row">
                                <div className="col">
                                    <MonthPicker year={currentDate.year} month={currentDate.month}
                                        onChange={this.changeDate} />
                                </div>
                                <div className="col">
                                    <TotalPrice income={totalIncome}
                                        outcome={totalOutcome} />
                                </div>
                            </div>
                        </header>
                        <div className="content-area py-3 px-3">
                        {isLoading &&
                            <Loader/>
                        }
                        {!isLoading && 
                            <React.Fragment>
                                <Tabs onTabChange={this.changeView} activeIndex={tabIndex}>
                                <Tab>
                                    <Ionicon
                                        className="rounded-circle mr-2"
                                        fontSize="25px"
                                        color="#007bff"
                                        icon="md-paper">
                                    </Ionicon>
                                    List Mode
                                </Tab>
                                <Tab>
                                    <Ionicon
                                        className="rounded-circle mr-2"
                                        fontSize="25px"
                                        color="#007bff"
                                        icon="md-pie">
                                    </Ionicon>
                                    Chart Mode
                                </Tab>
                                <Tab>
                                    <Ionicon
                                        className="rounded-circle mr-2"
                                        fontSize="25px"
                                        color="#007bff"
                                        icon="md-analytics">
                                    </Ionicon>
                                    Weekly Trend
                                </Tab>
                            </Tabs>

                            <CreateButton onCreateItem={this.createItem} />
                            {tabMode === LIST_MODE &&
                                <PriceList items={itemswithCategory} onEditItem={this.modeifyItem}
                                    onDeleteItem={this.deleteItem} />
                            }

                            {tabMode === CHART_MODE &&
                                <React.Fragment>
                                <PieFigure title={`Outcome Chart for ${currentDate.year}-${padZeroToLeft(currentDate.month)}`} items={filterItemsByType(itemswithCategory,TYPE_OUTCOME)}></PieFigure>
                                <PieFigure title={`Income Chart for ${currentDate.year}-${padZeroToLeft(currentDate.month)}`} items={filterItemsByType(itemswithCategory, TYPE_INCOME)}></PieFigure>
                                </React.Fragment>
                            }

                            {tabMode === TREND_MODE &&
                                <React.Fragment>
                                <LineFigure title={`Weekly Trend for ${currentDate.year}-${padZeroToLeft(currentDate.month)}`} items={itemswithCategory}></LineFigure>  
                                </React.Fragment>
                            }

                            </React.Fragment>
                        }
                        </div>
                    <Footer />
                </React.Fragment>
                
            )
    }
}
export default withRouter(withContext(Home));

//<Carousels height={720} width={'100%'} totalItems={images.length} items={images}>
//</Carousels>
//<div className="row mb-5 justify-content-center">
//<img src={logo} className="App-logo" alt="logo" />
//                            </div >

{//headLines && headLines.length !== 0 &&
    //<Carousels height={360} width={'75%'} totalItems={headLines.length} items={headLines}>
    //</Carousels>
}