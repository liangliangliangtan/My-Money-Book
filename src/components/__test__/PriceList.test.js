import React from 'react'
import { shallow } from 'enzyme'
import PriceList from '../PriceList'
import Ionicon from 'react-ionicons';
import {items,categories} from '../../containers/Home'
import toJson from 'enzyme-to-json';


/**
 * Shallow rendering is useful to constrain yourself to testing a component as a unit, 
 * and to ensure that your tests aren't indirectly asserting on behavior of child components.
 */
const itemswithCategory = items.map((item) => {
    item.category = categories[item.categoryId];
    return item;
})

const props = {
    items: itemswithCategory,
    onEditItem: jest.fn(), // jest mock function 
    onDeleteItem: jest.fn(),
}

let wrapper; 
describe('test PriceList component', () => {
    
    beforeEach(() => {
        wrapper = shallow(<PriceList {...props}/>)
    })
    
    /**
     * 1. generate component snapshot 
     */
    it('should render the component to match snapshort',()=>{
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    
    /**
     * 2. Test PriceList component render all items
     */
    it('should render correct price items length', ()ã€€=>{
        expect(wrapper.find('.list-group-item').length).toEqual(itemswithCategory.length)
    })

    /**
     * 3. Test each item 's icon was correctly rendered 
     */
    it('should render correct icon and price for each item', () =>{
        //find all Ionicon for the first item 
        const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
        // Should have 3 icons 
        expect(iconList.length).toEqual(3)
        // Should be the correct iconName
        expect(iconList.first().props().icon).toEqual(itemswithCategory[0].category.iconName)
    })
    
    /**
     * 4. Test callback function ,simulate click event
     */
    it('should trigger the correct function call back',() =>{
        const firstItem = wrapper.find('.list-group-item').first();
        //edit event  
        firstItem.find('a').first().simulate('click', { preventDefault(){}})
        expect(props.onEditItem).toHaveBeenCalledWith(itemswithCategory[0])
        //delete  event
        firstItem.find('a').last().simulate('click', {preventDefault(){}})
        expect(props.onDeleteItem).toHaveBeenCalledWith(itemswithCategory[0])
    })

})
