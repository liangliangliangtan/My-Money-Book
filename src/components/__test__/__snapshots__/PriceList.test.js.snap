// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshort 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item d-flex 
 justify-content-between align-items-center"
    key="1"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="30px"
        icon="md-restaurant"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#347eff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      testItem
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
       
      200
    </span>
    <span
      className="col-2"
    >
      2018-09-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="20px"
        icon="md-create"
        rotate={false}
        shake={false}
        style={Object {}}
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="20px"
        icon="md-trash"
        rotate={false}
        shake={false}
        style={Object {}}
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex 
 justify-content-between align-items-center"
    key="2"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="30px"
        icon="md-train"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#347eff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      testItem2
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      +
       
      300
    </span>
    <span
      className="col-2"
    >
      2018-09-10
    </span>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="20px"
        icon="md-create"
        rotate={false}
        shake={false}
        style={Object {}}
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <Ionicon
        beat={false}
        color="#6BE61A"
        fontSize="20px"
        icon="md-trash"
        rotate={false}
        shake={false}
        style={Object {}}
      />
    </a>
  </li>
</ul>
`;
